src/ui/auth/rackspace/rackspace.go

```diff
@@ -0,0 +1,174 @@
+		return nil, err
+	}
+
+	// check for auth server error
+	if resp.StatusCode >= http.StatusInternalServerError {
+		log.Errorf("ProvidedUsername=%s Error=%s", m.Principal, authRespBody)
+		return nil, fmt.Errorf("%d %s", resp.StatusCode, authRespBody)
+	}
+
+	// read auth response body as json
+	authResp := AuthResponse{}
+	err = json.Unmarshal([]byte(authRespBody), &authResp)
+	if err != nil {
+		log.Errorf("ProvidedUsername=%s Error=%v", m.Principal, err)
+		return nil, err
+	}
+
+	log.Debugf("ProvidedUsername=%s Authenticated=%t", m.Principal, authResp.Status.Authenticated)
+
+	// check for any status other than OK
+	if resp.StatusCode != http.StatusOK {
+		log.Errorf("ProvidedUsername=%s Error=%s", m.Principal, authResp.Status.Error)
+		return nil, fmt.Errorf("%d %s", resp.StatusCode, authResp.Status.Error)
+	}
+
+	log.Debugf("ProvidedUsername=%s UID=%s BackendUsername=%s", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username)
+
+	// check if the user already exists in the Harbor DB. if the user doesn't exist, create it.
+	user := models.User{}
+	// set the Harbor Realname to the kubernetes-auth backend's UID because the UID is a static ID
+	// whereas the kubernetes-auth backend's Username can change (so put it in the Harbor Username field for convenience)
+	user.Realname = authResp.Status.User.UID
+
+	exist, err := dao.UserExists(user, "realname")
+	if err != nil {
+		log.Errorf("ProvidedUsername=%s UID=%s BackendUsername=%s Error=%v", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username, err)
+		return nil, err
+	}
+
+	if exist {
+		log.Debugf("ProvidedUsername=%s UID=%s BackendUsername=%s exists in Harbor DB", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username)
+
+		existingUser, err := dao.GetUser(user)
+		if err != nil {
+			log.Errorf("ProvidedUsername=%s UID=%s BackendUsername=%s Error=%v", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username, err)
+			return nil, err
+		}
+
+		// if the username changed in kubernetes-auth backend, update it in the Harbor DB
+		if existingUser.Username != authResp.Status.User.Username {
+			log.Debugf("ProvidedUsername=%s UID=%s BackendUsername=%s backend username change so updating Harbor DB", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username)
+
+			user.Username = authResp.Status.User.Username
+			existingUser.Username = authResp.Status.User.Username
+
+			err = dao.ChangeUserProfile(user)
+			if err != nil {
+				log.Errorf("ProvidedUsername=%s UID=%s BackendUsername=%s Error=%v", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username, err)
+				return nil, err
+			}
+		}
+
+		user = *existingUser
+	} else {
+		log.Debugf("ProvidedUsername=%s UID=%s BackendUsername=%s does not exist in Harbor DB so creating new user", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username)
+
+		user.Username = authResp.Status.User.Username
+		user.Password = "ThisPassw0rdIsNotused"
```

Comment:
this will generate a random 32 char string:
```
func RandString() string {
    letterBytes := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    b := make([]byte, 32)
    for i := range b {
        b[i] = letterBytes[rand.Intn(len(letterBytes))]
    }
    return string(b)
}
```